### PR TITLE
Fix type annotations (from Samsung, AI Center, Cambridge)

### PR DESCRIPTION
**Pull request by rogiervd**
## What does this PR do?

Fix existing type annotations. These are meant to be uncontroversial.

There are a few types of categories:
* Ungrammatical or mis-spelled type annotations.
* Arguments `argument: int = None` where the type should be `Optional[int]`
* The opposite: arguments `argument: Optional[int] = 3072` where the type should be `int`, because `Optional` means "may be `None`".
* Various other cases.

These were flagged by Pyright (see https://github.com/speechbrain/speechbrain/pull/2901).

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/utils/logger.py [hunk 1]
```diff
@@ -128,7 +128,7 @@ def log(self, level: int, msg: str, *args: tuple, **kwargs: dict):
                 self.logger.log(level, msg, *args, **kwargs)
 
     @functools.lru_cache(None)
-    def warning_once(self, *args: tuple, **kwargs: dict):
+    def warning_once(self, *args, **kwargs):
```
Comment (rogiervd) [inline speechbrain/utils/logger.py:131]: From https://typing.python.org/en/latest/spec/callables.html#annotating-args-and-kwargs:
> Therefore, the definition:
```
def func(*args: str, **kwargs: int): ...
```
> means that the function accepts an arbitrary number of positional arguments of type str and an arbitrary number of keyword arguments of type int.

File: speechbrain/dataio/sampler.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 from collections import Counter
 from operator import itemgetter
-from typing import List
+from typing import List, Optional, Union
```
Comment (pplantinga) [inline speechbrain/dataio/sampler.py:16]: Should `List` be removed here?

File: speechbrain/lobes/models/transformer/Transformer.py [hunk 3]
```diff
@@ -104,26 +104,26 @@ def __init__(
         num_decoder_layers=6,
         d_ffn=2048,
         dropout=0.1,
-        activation=nn.ReLU,
+        activation: type = nn.ReLU,
```
Comment (pplantinga) [inline speechbrain/lobes/models/transformer/Transformer.py:107]: Should we use `Type[nn.Module]` here?

File: speechbrain/nnet/RNN.py [hunk 4]
```diff
@@ -1914,8 +1916,9 @@ def __init__(
         if self.output_gate:
             self.o_gate = nn.Sigmoid()
 
-    def forgetMult(self, f, x, hidden):
-        # type: (Tensor, torch.Tensor, Optional[Tensor]) -> torch.Tensor # noqa F821
+    def forgetMult(
+        self, f: torch.Tensor, x: torch.Tensor, hidden: Optional[torch.Tensor]
```
Comment (pplantinga) [inline speechbrain/nnet/RNN.py:1920]: Should this be `hidden: Optional[torch.Tensor] = None` ?

**Review by pplantinga — COMMENTED** [review]
Overall, looks good. Just a few questions here

File: speechbrain/dataio/sampler.py (hunk 2 above)
Comment (rogiervd) [reply speechbrain/dataio/sampler.py:16]: There are existing uses of `List` which I have not changed, but we could/should do this in another PR, now that python>=3.9.

File: speechbrain/lobes/models/transformer/Transformer.py (hunk 3 above)
Comment (rogiervd) [reply speechbrain/lobes/models/transformer/Transformer.py:107]: Yes but I can't get it to work. I may be missing something.

```
speechbrain/lobes/models/transformer/Transformer.py:228:28 - error: Argument of type "type[Module]" cannot be assigned to parameter "activation" of type "type[ReLU]" in function "__init__"
    "type[Module]" is not assignable to "type[ReLU]"
    Type "type[Module]" is not assignable to type "type[ReLU]" (reportArgumentType)
```

File: speechbrain/nnet/RNN.py (hunk 4 above)
Comment (rogiervd) [reply speechbrain/nnet/RNN.py:1920]: No default value was given in the original, and I haven't changed the behaviour. This may have been the original intention, or not. Of course `Optional[int]` is `int | None`: `Optional` does not mean that the argument is optional.

File: speechbrain/lobes/models/transformer/Transformer.py (hunk 3 above)
Comment (pplantinga) [reply speechbrain/lobes/models/transformer/Transformer.py:107]: I messed around with this a bit and still couldn't get it to work, I'm fine with `type` for now until we find something better.

**Review by pplantinga — APPROVED** [review]
LGTM
